aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+    linalg_eig_make_complex_eigenvectors_impl<scalar_t>(complex_vectors, complex_values, real_vectors);
+  });
+  return complex_vectors;
+}
+
+DEFINE_DISPATCH(linalg_eig_stub);
+
+std::tuple<Tensor&, Tensor&> linalg_eig_out_info(const Tensor& input, Tensor& values, Tensor& vectors, Tensor& infos, bool compute_eigenvectors) {
+  // MAGMA doesn't have GPU interface for GEEV routine, it requires inputs to be on CPU
+  // therefore we create all intermediate tensors on CPU
+  auto options = input.options().device(at::kCPU);
+
+  // These internal asserts make explicit the assumptions in the implementation
+  // Error check with the actual error messages are done on the higher level of the hierarchy of calls
+  TORCH_INTERNAL_ASSERT(input.dim() >= 2);
+  TORCH_INTERNAL_ASSERT(input.size(-2) == input.size(-1));
+
+  // for real-valued 'input', eigenvalues can be real-valued or complex-valued
+  TORCH_INTERNAL_ASSERT((toComplexType(input.scalar_type()) == values.scalar_type()) || (input.scalar_type() == values.scalar_type()));
+  TORCH_INTERNAL_ASSERT(values.device() == at::kCPU);
+
+  // for real-valued 'input', eigenvectors can be real-valued or complex-valued
+  if (compute_eigenvectors) {
+    TORCH_INTERNAL_ASSERT((toComplexType(input.scalar_type()) == vectors.scalar_type()) || (input.scalar_type() == vectors.scalar_type()));
+    TORCH_INTERNAL_ASSERT(vectors.device() == at::kCPU);
+  }
+
+  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.device() == at::kCPU);
+  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
+  TORCH_INTERNAL_ASSERT(infos.is_contiguous());
+
+  // if 'vectors' has no elements we can modify it
+  if (vectors.numel() == 0 && compute_eigenvectors) {
+    vectors.resize_(input.sizes(), MemoryFormat::Contiguous);
+    vectors.transpose_(-2, -1);  // make 'vectors' to have Fortran contiguous memory layout
+  }
+
+  // if 'values' has no elements we can modify it
+  auto values_shape = IntArrayRef(input.sizes().data(), input.dim()-1);  // input.shape[:-1]
+  if (values.numel() == 0) {
+    values.resize_(values_shape, MemoryFormat::Contiguous);
+  }
+
+  // 'vectors' must be in batched column major order (Fortran contiguous)
+  if (compute_eigenvectors) {
+    TORCH_INTERNAL_ASSERT(vectors.transpose(-2, -1).is_contiguous());
+    TORCH_INTERNAL_ASSERT(vectors.sizes().equals(input.sizes()));
+  }
+
+  // 'values' must be contiguous
+  TORCH_INTERNAL_ASSERT(values.is_contiguous());
+  TORCH_INTERNAL_ASSERT(values.sizes().equals(values_shape));
+
+  // if 'input' is complex then use 'values' directly else create a temporary to hold the real and imaginary parts
+  // and then use at::complex_out
+  Tensor real_imag_values = values;
+
+  // if 'input' is complex then use 'vectors' directly else maybe create a temporary to hold real vectors
+  // and then use linalg_eig_make_complex_eigenvectors
+  Tensor maybe_complex_vectors = vectors;
+  if (!input.is_complex()) {
```

Comment:
This is a lot of work for inputs that aren't complex.

Would you elaborate on what's going on with floating point vs complex inputs? Does it make sense to cast all inputs to complex?

Reply:
Casting all inputs to complex dtype would make LAPACK perform unnecessarily worse. This post-processing is required for inputs that aren't complex just because LAPACK gives output in compressed form using only real-valued arrays.
I will profile the implementation to show this is actually not a lot of work computationally.

Reply:
For a matrix of size 4096x4096 `linalg_eig_make_complex_eigenvectors` takes ~0.16% of the total execution time of `linalg_eig_out_info`, ~99.84% is spent in `linalg_eig_stub`.
For a matrix of size 256x256 `linalg_eig_make_complex_eigenvectors` takes ~0.6%.

Reply:
OK.